controllers/k8ssandra/cassandra_telemetry_reconciler.go

```diff
@@ -0,0 +1,94 @@
+// Logic in this file reconciles telemetry resources into the state declared in the CRs.
+
+package k8ssandra
+
+import (
+	"context"
+	"errors"
+	"github.com/go-logr/logr"
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+	api "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	k8ssandraapi "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/result"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/telemetry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *K8ssandraClusterReconciler) reconcileCassandraDCTelemetry(
+	ctx context.Context,
+	kc *api.K8ssandraCluster,
+	dcTemplate api.CassandraDatacenterTemplate,
+	actualDc *cassdcapi.CassandraDatacenter,
+	logger logr.Logger,
+	remoteClient client.Client,
+) result.ReconcileResult {
+	logger.Info("reconciling telemetry")
+	mergedSpec := kc.Spec.Cassandra.CassandraTelemetry.Merge(dcTemplate.CassandraTelemetry)
+	cfg := telemetry.PrometheusResourcer{
+		MonitoringTargetNS:   actualDc.Namespace,
+		MonitoringTargetName: actualDc.Name,
+		ServiceMonitorName:   kc.Name + "-" + actualDc.Name + "-" + "cass-servicemonitor",
+		Logger:               logger,
+		CommonLabels:         mustLabels(kc.Name, actualDc.Name),
+	}
+	logger.Info("merged TelemetrySpec constructed", "mergedSpec", mergedSpec, "cluster", kc.Name)
+	// Confirm telemetry config is valid (e.g. Prometheus is installed if it is requested.)
+	promInstalled, err := telemetry.IsPromInstalled(remoteClient, logger)
+	if err != nil {
+		return result.Error(err)
+	}
+	validConfig, err := telemetry.SpecIsValid(mergedSpec, promInstalled)
+	if err != nil {
+		return result.Error(errors.New("could not determine if telemetry config is valid"))
+	}
+	if !validConfig {
+		return result.Error(errors.New("telemetry spec was invalid for this cluster - is Prometheus installed if you have requested it"))
+	}
+	// If Prometheus not installed bail here.
+	if !promInstalled {
+		return result.Continue()
+	}
+	// Determine if we want a cleanup or a resource update.
+	switch {
+	case mergedSpec == nil:
+		logger.Info("Telemetry not present for CassDC, will delete resources", "mergedSpec", mergedSpec)
+		if err := cfg.CleanupResources(ctx, remoteClient); err != nil {
+			return result.Error(err)
+		}
+		return result.Continue()
+	case mergedSpec.Prometheus == nil:
+		logger.Info("Telemetry not present for CassDC, will delete resources", "mergedSpec", mergedSpec)
+		if err := cfg.CleanupResources(ctx, remoteClient); err != nil {
```

Comment:
`CleanupResources` uses `DeleteAllOf`,  but will there be scenarios in which we have more than one ServiceMonitor for a CassandraDatacenter? If not then we look for the single ServiceMonitor (if it exists) to be deleted. We can perform the lookup easily enough since the ServiceMonitor name is basically pre-computed. Note that if we could not easily compute the ServiceMontior name, then we should be able to query for the ServiceMonitor using a label selector.

Reply:
You [asked for ](https://github.com/k8ssandra/k8ssandra-operator/pull/241#discussion_r766950899)this change earlier? 

In either event, I prefer to use `DeleteAllOf` just in case we eventually allow the user to set their own name on the `ServiceMonitor`, it seems more idiomatic to work with labels than names.

Reply:
On my first review, I only noticed the method deleting objects could potentially be simplified by using `DeleteAllOf` On my subsequent review(s) when I had a better understanding of the code it occurred to me that `DeleteAllOf` may not even necessary if we are only dealing with a single ServiceMonitor. Apologies for any confusion that I caused.

> I prefer to use DeleteAllOf just in case we eventually allow the user to set their own name on the ServiceMonitor

We will know the ServiceMonitor name whether it is computed or it was specified by the user, right? If specified, it would be included in the spec. More importantly, using `DeleteAllOf` increases security surface area as we have to add a permission that really isn't necessary. 

>it seems more idiomatic to work with labels than names

We delete objects by key (i.e., types.NamespacedName / client.ObjectKey) pretty much everywhere except for a couple places in tests. I think it's the same for cass-operator.

Reply:
I don't think it makes a difference to any hypothetical attack surface given that an attacker can list and individually delete resources as easily as running DeleteAllOf. It is odd to me that these are even distinct permissions in the RBAC. 

Regardless, if you'd prefer it implemented as a single deletion then I'll do so - it will make no difference to the behaviour of the function.

Reply:
When I started to implement this some complications came up around the assumptions inherent in the `client.Get(ctx, dTarget)... client.Delete(ctx, dTarget)` sequence. It assumes that the deletion target definitely exists, and that assumption is not concordant with our use case here (cleanup is called whenever the telemetry is not explicitly enabled, even if no telemetry resources actually exist).

Not moving forward with this because the current usage of `DeleteAllOf` seems more appropriate for our use case.